pkg/controller/egress/validate.go

```diff
@@ -0,0 +1,166 @@
+)
+
+func (c *EgressController) ValidateExternalIPPool(review *admv1.AdmissionReview) *admv1.AdmissionResponse {
+	var result *metav1.Status
+	var msg string
+	allowed := true
+
+	klog.V(2).Info("Validating ExternalIPPool", "request", review.Request)
+	var newObj, oldObj crdv1alpha2.ExternalIPPool
+	if review.Request.Object.Raw != nil {
+		if err := json.Unmarshal(review.Request.Object.Raw, &newObj); err != nil {
+			klog.ErrorS(err, "Error de-serializing current ExternalIPPool")
+			return newAdmissionResponseForErr(err)
+		}
+	}
+	if review.Request.OldObject.Raw != nil {
+		if err := json.Unmarshal(review.Request.OldObject.Raw, &oldObj); err != nil {
+			klog.ErrorS(err, "Error de-serializing old ExternalIPPool")
+			return newAdmissionResponseForErr(err)
+		}
+	}
+
+	switch review.Request.Operation {
+	case admv1.Create:
+		klog.V(2).Info("Validating CREATE request for ExternalIPPool")
+		// Always allow CREATE request.
+	case admv1.Update:
+		// Tier priority updates are not allowed
+		klog.V(2).Info("Validating UPDATE request for ExternalIPPool")
+
+		oldIPRangeSet := getIPRangeSet(oldObj.Spec.IPRanges)
+		newIPRangeSet := getIPRangeSet(newObj.Spec.IPRanges)
+		deletedIPRanges := oldIPRangeSet.Difference(newIPRangeSet)
+		if deletedIPRanges.Len() > 0 {
+			allowed = false
+			msg = fmt.Sprintf("existing IPRanges %v cannot be deleted", deletedIPRanges.List())
+		}
+	case admv1.Delete:
+		klog.V(2).Info("Validating DELETE request for ExternalIPPool")
+		// Always allow DELETE request.
+	}
+
+	if msg != "" {
+		result = &metav1.Status{
+			Message: msg,
+		}
+	}
+	return &admv1.AdmissionResponse{
+		Allowed: allowed,
+		Result:  result,
+	}
+}
+
+func (c *EgressController) ValidateEgress(review *admv1.AdmissionReview) *admv1.AdmissionResponse {
+	var result *metav1.Status
+	var msg string
+	allowed := true
+
+	klog.V(2).Info("Validating Egress", "request", review.Request)
+	var newObj, oldObj crdv1alpha2.Egress
+	if review.Request.Object.Raw != nil {
+		if err := json.Unmarshal(review.Request.Object.Raw, &newObj); err != nil {
+			klog.ErrorS(err, "Error de-serializing current Egress")
+			return newAdmissionResponseForErr(err)
+		}
+	}
+	if review.Request.OldObject.Raw != nil {
+		if err := json.Unmarshal(review.Request.OldObject.Raw, &oldObj); err != nil {
+			klog.ErrorS(err, "Error de-serializing old Egress")
+			return newAdmissionResponseForErr(err)
+		}
+	}
+
+	shouldAllow := func(oldEgress, newEgress *crdv1alpha2.Egress) (bool, string) {
+		// Allow it if EgressIP and ExternalIPPool don't change.
+		if newEgress.Spec.EgressIP == oldEgress.Spec.EgressIP && newEgress.Spec.ExternalIPPool == oldEgress.Spec.ExternalIPPool {
+			return true, ""
+		}
+		// Only validate whether the specified Egress IP is in the Pool when they are both set.
+		if newEgress.Spec.EgressIP == "" || newEgress.Spec.ExternalIPPool == "" {
+			return true, ""
+		}
+		ip := net.ParseIP(newEgress.Spec.EgressIP)
+		if ip == nil {
```

Comment:
can we use openAPI to validate that EgressIP is always an IP?

Reply:
It has been validated in CRD schema, here it needs to be converted to net.IP anyway, so just check it by the way.